dali/python/nvidia/dali/plugin/pytorch/experimental/proxy/__init__.py

```diff
@@ -0,0 +1,472 @@
+        is_scheduled is False)
+        """
+        self.batch_id = batch_id
+        self.is_scheduled = is_scheduled
+        self.inputs = inputs
+
+
+class DALIProcessedSampleRef:
+    """
+    Placeholder for a pipeline run reference, which is returned by the data worker instead of
+    the actual data. The PyTorch worker returns this trivial object, only containing information
+    about this proxy instance and the input data to the pipeline. Later in the collate function,
+    we send the data for execution to DALI.
+    """
+
+    def __init__(self, proxy, inputs):
+        self.proxy = proxy
+        self.inputs = inputs
+        if len(inputs) != len(self.proxy.dali_input_names):
+            raise RuntimeError(
+                f"Unexpected number of inputs. Expected: {self.dali_input_names}, got: {inputs}"
+            )
+
+
+class _DALIProxy:
+    def __init__(self, dali_input_names, dali_input_q, deterministic):
+        # External source instance names
+        self.dali_input_names = dali_input_names
+        # If True, the inputs are scheduled after the data loader has returned
+        self.deterministic = deterministic
+        # Shared queue with the server
+        self.dali_input_q = dali_input_q
+        # Torch worker id, to be filled on first call to worker_id()
+        self._worker_id = None
+        # Iteration index for the current worker
+        self.data_idx = 0
+
+    @property
+    def worker_id(self):
+        """
+        Getter for 'worker_id'. In case of torch data worker it is the worker info, and
+        in case of a thread the thread identifier
+        """
+        if self._worker_id is None:
+            worker_info = torchdata.get_worker_info()
+            self._worker_id = worker_info.id if worker_info else threading.get_ident()
+        return self._worker_id
+
+    def _schedule_batch(self, inputs):
+        # Identifier of this request
+        batch_id = (self.worker_id, self.data_idx)
+        self.data_idx = self.data_idx + 1
+        if not self.deterministic:
+            torch.cuda.nvtx.range_push(f"dali_proxy.dali_input_q.put {batch_id}")
+            self.dali_input_q.put((batch_id, inputs))
+            torch.cuda.nvtx.range_pop()
+            return DALIPipelineRunRef(batch_id, is_scheduled=True)
+        else:
+            return DALIPipelineRunRef(batch_id, is_scheduled=False, inputs=inputs)
+
+    def __call__(self, *inputs):
+        """
+        Returns a reference to the pipeline run
+        """
+        if len(inputs) != len(self.dali_input_names):
+            raise RuntimeError(
+                f"Unexpected number of inputs. Expected: {self.dali_input_names}, got: {inputs}"
+            )
+        return DALIProcessedSampleRef(self, inputs)
+
+
+class DALIServer:
+    def __init__(self, pipeline, input_names=None, deterministic=False):
+        """
+        Initializes a new DALI server instance.
```

Comment:
It's good to see docs with examples! If we want the proxy to be usable though, we definitely need to provide more context and more details here. This can be done in a follow-up PR, of course - this one is already quite big.